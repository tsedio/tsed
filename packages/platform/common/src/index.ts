/**
 * @file Automatically generated by @tsed/barrels.
 */
export * from "./builder/PlatformBuilder.js";
export * from "./config/interfaces/PlatformJsonMapperSettings.js";
export * from "./config/interfaces/PlatformLoggerSettings.js";
export * from "./config/interfaces/PlatformMulterSettings.js";
export * from "./config/interfaces/PlatformRouterSettings.js";
export * from "./config/interfaces/PlatformStaticsSettings.js";
export * from "./config/services/PlatformConfiguration.js";
export * from "./constants/routerOptions.js";
export * from "./decorators/multer/multerFileSize.js";
export * from "./decorators/multer/multerOptions.js";
export * from "./decorators/multer/multipartFile.js";
export * from "./decorators/params/error.js";
export * from "./decorators/params/next.js";
export * from "./decorators/params/request.js";
export * from "./decorators/params/response.js";
export * from "./decorators/PlatformProvider.js";
export * from "./domain/AnyToPromiseWithCtx.js";
export * from "./domain/EndpointMetadata.js";
export * from "./domain/PlatformContext.js";
export * from "./domain/PlatformRouteDetails.js";
export * from "./exports.js";
export * from "./interfaces/AfterInit.js";
export * from "./interfaces/AfterListen.js";
export * from "./interfaces/AfterRoutesInit.js";
export * from "./interfaces/BeforeInit.js";
export * from "./interfaces/BeforeListen.js";
export * from "./interfaces/BeforeRoutesInit.js";
export * from "./interfaces/HandlerType.js";
export * from "./interfaces/IncomingEvent.js";
export * from "./interfaces/OnReady.js";
export * from "./interfaces/OnRequest.js";
export * from "./interfaces/OnResponse.js";
export * from "./interfaces/OnRoutesInit.js";
export * from "./interfaces/PlatformRouteOptions.js";
export * from "./interfaces/ResponseErrorObject.js";
export * from "./interfaces/Route.js";
export * from "./middlewares/PlatformAcceptMimesMiddleware.js";
export * from "./middlewares/PlatformMulterMiddleware.js";
export * from "./services/FakeAdapter.js";
export * from "./services/Platform.js";
export * from "./services/PlatformAdapter.js";
export * from "./services/PlatformApplication.js";
export * from "./services/PlatformHandler.js";
export * from "./services/PlatformMiddlewaresChain.js";
export * from "./services/PlatformRequest.js";
export * from "./services/PlatformResponse.js";
export * from "./services/PlatformTest.js";
export * from "./utils/closeServer.js";
export * from "./utils/createContext.js";
export * from "./utils/createHttpServer.js";
export * from "./utils/createHttpsServer.js";
export * from "./utils/createInjector.js";
export * from "./utils/createServer.js";
export * from "./utils/ensureContext.js";
export * from "./utils/getStaticsOptions.js";
export * from "./utils/listenServer.js";
export * from "./utils/mapReturnedResponse.js";
export * from "./utils/printRoutes.js";
export * from "./utils/registerPlatformAdapter.js";
export * from "./utils/setResponseHeaders.js";
