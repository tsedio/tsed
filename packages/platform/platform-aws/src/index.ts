export * from "./components/PlatformAws";
export * from "./decorators/awsContext";
export * from "./decorators/awsEvent";
