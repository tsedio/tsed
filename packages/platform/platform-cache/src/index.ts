export * from "./interfaces";
export * from "./services/PlatformCache";
export * from "./decorators/useCache";
export * from "./PlatformCacheModule";
