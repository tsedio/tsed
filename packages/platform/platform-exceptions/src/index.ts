// components
export * from "./components/StringErrorFilter";
export * from "./components/ExceptionFilter";
export * from "./components/ErrorFilter";
export * from "./components/MongooseErrorFilter";

// decorators
export * from "./decorators/catch";

// domain
export * from "./domain/ExceptionFiltersContainer";
export * from "./domain/ExceptionSchema";
export * from "./interfaces/ExceptionFilterMethods";

// middlewares
export * from "./services/PlatformExceptions";

// errors
export * from "./errors/ResourceNotFound";
