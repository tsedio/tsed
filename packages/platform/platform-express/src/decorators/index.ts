export * from "./routerSettings";
export * from "./caseSensitive";
export * from "./mergeParams";
export * from "./strict";
