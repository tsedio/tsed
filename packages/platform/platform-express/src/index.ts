/**
 * @file Automatically generated by barrelsby.
 */

export * from "./components/PlatformExpress";
export * from "./interfaces/PlatformExpressSettings";
export * from "./interfaces/PlatformExpressStaticsOptions";
export * from "./interfaces/interfaces";
export * from "./middlewares/multerMiddleware";
export * from "./middlewares/staticsMiddleware";
