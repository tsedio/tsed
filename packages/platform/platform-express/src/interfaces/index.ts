import Express from "express";
import {PlatformExpressSettings} from "./PlatformExpressSettings";

export * from "./PlatformExpressSettings";
export * from "./PlatformExpressStaticsOptions";

declare global {
  namespace TsED {
    export interface Configuration {
      /**
       * Configuration related to Express platform application.
       */
      express: PlatformExpressSettings;
    }

    export interface NextFunction extends Express.NextFunction {}
  }
}
