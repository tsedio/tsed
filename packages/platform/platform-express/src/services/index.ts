export * from "./PlatformExpressApplication";
export * from "./PlatformExpressHandler";
export * from "./PlatformExpressResponse";
export * from "./PlatformExpressRequest";
export * from "./PlatformExpressRouter";
