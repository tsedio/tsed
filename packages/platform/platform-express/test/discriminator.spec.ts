import "@tsed/ajv";
import {BodyParams, Controller, Patch, PlatformTest, Post} from "@tsed/common";
import {PlatformTestSdk} from "@tsed/platform-test-sdk";
import {DiscriminatorKey, DiscriminatorValue, OneOf, Partial, Property, Required, Returns} from "@tsed/schema";
import "@tsed/swagger";
import SuperTest from "supertest";
import {PlatformExpress} from "../src/index";
import {rootDir, Server} from "./app/Server";

class Event {
  @DiscriminatorKey() // declare this property as discriminator key
  type: string;

  @Property()
  value: string;
}

@DiscriminatorValue("page_view") // or @DiscriminatorValue() value can be inferred by the class name
class PageView extends Event {
  @Required()
  url: string;
}

@DiscriminatorValue("action", "click_action")
class Action extends Event {
  @Required()
  event: string;
}

@DiscriminatorValue()
class CustomAction extends Event {
  @Required()
  event: string;

  @Property()
  meta: string;
}

type EventTypes = PageView | Action | CustomAction;

@Controller("/discriminator")
class TestDiscriminator {
  @Post("/scenario-1")
  @Returns(200).OneOf(Event)
  scenario1(@BodyParams() @OneOf(Event) event: EventTypes) {
    expect(event).toBeInstanceOf(PageView);
    return event;
  }

  @Post("/scenario-2")
  @Returns(200, Array).OneOf(Event)
  scenario2(@BodyParams() @OneOf(Event) events: EventTypes[]) {
    expect(events[0]).toBeInstanceOf(PageView);
    expect(events[1]).toBeInstanceOf(Action);
    expect(events[2]).toBeInstanceOf(Action);
    expect(events[3]).toBeInstanceOf(CustomAction);
    return events;
  }

  @Patch("/scenario-3")
  @Returns(200).OneOf(Event)
  scenario3(@BodyParams() @OneOf(Event) @Partial() event: EventTypes) {
    expect(event).toBeInstanceOf(PageView);
    return event;
  }
}

const utils = PlatformTestSdk.create({
  rootDir,
  platform: PlatformExpress,
  server: Server
});
describe("Discriminator", () => {
  let request: SuperTest.Agent;

  beforeAll(
    utils.bootstrap({
      mount: {
        "/rest": [TestDiscriminator]
      },
      swagger: [
        {
          path: "/docs",
          specVersion: "3.0.1"
        }
      ]
    })
  );
  afterAll(utils.reset);

  beforeAll(() => {
    request = SuperTest(PlatformTest.callback());
  });

  describe("os3", () => {
    it("should generate the spec", async () => {
      const {body} = await request.get("/docs/swagger.json");
      expect(body).toMatchSnapshot();
    });
  });

  describe("POST /rest/discriminator: scenario 1", () => {
    it("should map correctly the data", async () => {
      const {body} = await request
        .post("/rest/discriminator/scenario-1")
        .send({
          type: "page_view",
          value: "value",
          url: "https://url.com"
        })
        .expect(200);

      expect(body).toEqual({
        type: "page_view",
        url: "https://url.com",
        value: "value"
      });
    });
    it("should throw an error when the type doesn't exists", async () => {
      const {body} = await request
        .post("/rest/discriminator/scenario-1")
        .send({
          type: "test",
          value: "value",
          url: "https://url.com"
        })
        .expect(400);

      expect(body).toEqual({
        errors: [
          {
            requestPath: "body",
            data: {
              type: "test",
              url: "https://url.com",
              value: "value"
            },
            dataPath: "",
            instancePath: "",
            keyword: "discriminator",
            message: 'value of tag "type" must be in oneOf',
            params: {
              error: "mapping",
              tag: "type",
              tagValue: "test"
            },
            schemaPath: "#/discriminator"
          }
        ],
        message:
          'Bad request on parameter "request.body".\nValue value of tag "type" must be in oneOf. Given value: {"type":"test","value":"value","url":"https://url.com"}',
        name: "AJV_VALIDATION_ERROR",
        status: 400
      });
    });
  });

  describe("POST /rest/discriminator: scenario 2", () => {
    it("should map correctly the data", async () => {
      const result = await request
        .post("/rest/discriminator/scenario-2")
        .send([
          {
            type: "page_view",
            value: "value",
            url: "https://url"
          },
          {
            type: "action",
            value: "value",
            event: "event"
          },
          {
            type: "click_action",
            value: "value",
            event: "event"
          },
          {
            type: "custom_action",
            value: "custom",
            event: "event",
            meta: "meta"
          }
        ])
        .expect(200);

      expect(result.body).toEqual([
        {
          type: "page_view",
          url: "https://url",
          value: "value"
        },
        {
          event: "event",
          type: "action",
          value: "value"
        },
        {
          event: "event",
          type: "click_action",
          value: "value"
        },
        {
          event: "event",
          meta: "meta",
          type: "custom_action",
          value: "custom"
        }
      ]);
    });
  });

  describe("PATCH /rest/discriminator: scenario 3", () => {
    it("should map correctly the data", async () => {
      const result = await request.patch("/rest/discriminator/scenario-3").send({
        type: "page_view",
        value: "value",
        url: "https://url"
      });
      expect(result.body).toEqual({type: "page_view", url: "https://url", value: "value"});
    });
  });
});
