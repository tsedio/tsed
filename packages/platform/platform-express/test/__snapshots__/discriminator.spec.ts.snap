// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Discriminator > os3 > should generate the spec 1`] = `
{
  "components": {
    "schemas": {
      "Action": {
        "properties": {
          "event": {
            "minLength": 1,
            "type": "string",
          },
          "type": {
            "enum": [
              "action",
              "click_action",
            ],
            "example": "action",
            "type": "string",
          },
          "value": {
            "type": "string",
          },
        },
        "required": [
          "event",
        ],
        "type": "object",
      },
      "ActionPartial": {
        "properties": {
          "event": {
            "minLength": 1,
            "type": "string",
          },
          "type": {
            "enum": [
              "action",
              "click_action",
            ],
            "example": "action",
            "type": "string",
          },
          "value": {
            "type": "string",
          },
        },
        "type": "object",
      },
      "CustomAction": {
        "properties": {
          "event": {
            "minLength": 1,
            "type": "string",
          },
          "meta": {
            "type": "string",
          },
          "type": {
            "example": "custom_action",
            "type": "string",
          },
          "value": {
            "type": "string",
          },
        },
        "required": [
          "event",
        ],
        "type": "object",
      },
      "CustomActionPartial": {
        "properties": {
          "event": {
            "minLength": 1,
            "type": "string",
          },
          "meta": {
            "type": "string",
          },
          "type": {
            "example": "custom_action",
            "type": "string",
          },
          "value": {
            "type": "string",
          },
        },
        "type": "object",
      },
      "PageView": {
        "properties": {
          "type": {
            "example": "page_view",
            "type": "string",
          },
          "url": {
            "minLength": 1,
            "type": "string",
          },
          "value": {
            "type": "string",
          },
        },
        "required": [
          "url",
        ],
        "type": "object",
      },
      "PageViewPartial": {
        "properties": {
          "type": {
            "example": "page_view",
            "type": "string",
          },
          "url": {
            "minLength": 1,
            "type": "string",
          },
          "value": {
            "type": "string",
          },
        },
        "type": "object",
      },
    },
  },
  "info": {
    "title": "Api documentation",
    "version": "1.0.0",
  },
  "openapi": "3.0.1",
  "paths": {
    "/rest/discriminator/scenario-1": {
      "post": {
        "operationId": "testDiscriminatorScenario1",
        "parameters": [],
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "discriminator": {
                  "propertyName": "type",
                },
                "oneOf": [
                  {
                    "$ref": "#/components/schemas/PageView",
                  },
                  {
                    "$ref": "#/components/schemas/Action",
                  },
                  {
                    "$ref": "#/components/schemas/CustomAction",
                  },
                ],
                "required": [
                  "type",
                ],
              },
            },
          },
          "required": false,
        },
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "discriminator": {
                    "propertyName": "type",
                  },
                  "oneOf": [
                    {
                      "$ref": "#/components/schemas/PageView",
                    },
                    {
                      "$ref": "#/components/schemas/Action",
                    },
                    {
                      "$ref": "#/components/schemas/CustomAction",
                    },
                  ],
                  "required": [
                    "type",
                  ],
                },
              },
            },
            "description": "Success",
          },
        },
        "tags": [
          "TestDiscriminator",
        ],
      },
    },
    "/rest/discriminator/scenario-2": {
      "post": {
        "operationId": "testDiscriminatorScenario2",
        "parameters": [],
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "items": {
                  "discriminator": {
                    "propertyName": "type",
                  },
                  "oneOf": [
                    {
                      "$ref": "#/components/schemas/PageView",
                    },
                    {
                      "$ref": "#/components/schemas/Action",
                    },
                    {
                      "$ref": "#/components/schemas/CustomAction",
                    },
                  ],
                  "required": [
                    "type",
                  ],
                },
                "type": "array",
              },
            },
          },
          "required": false,
        },
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "items": {
                    "discriminator": {
                      "propertyName": "type",
                    },
                    "oneOf": [
                      {
                        "$ref": "#/components/schemas/PageView",
                      },
                      {
                        "$ref": "#/components/schemas/Action",
                      },
                      {
                        "$ref": "#/components/schemas/CustomAction",
                      },
                    ],
                    "required": [
                      "type",
                    ],
                  },
                  "type": "array",
                },
              },
            },
            "description": "Success",
          },
        },
        "tags": [
          "TestDiscriminator",
        ],
      },
    },
    "/rest/discriminator/scenario-3": {
      "patch": {
        "operationId": "testDiscriminatorScenario3",
        "parameters": [],
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "discriminator": {
                  "propertyName": "type",
                },
                "oneOf": [
                  {
                    "$ref": "#/components/schemas/PageViewPartial",
                  },
                  {
                    "$ref": "#/components/schemas/ActionPartial",
                  },
                  {
                    "$ref": "#/components/schemas/CustomActionPartial",
                  },
                ],
                "required": [
                  "type",
                ],
              },
            },
          },
          "required": false,
        },
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "discriminator": {
                    "propertyName": "type",
                  },
                  "oneOf": [
                    {
                      "$ref": "#/components/schemas/PageView",
                    },
                    {
                      "$ref": "#/components/schemas/Action",
                    },
                    {
                      "$ref": "#/components/schemas/CustomAction",
                    },
                  ],
                  "required": [
                    "type",
                  ],
                },
              },
            },
            "description": "Success",
          },
        },
        "tags": [
          "TestDiscriminator",
        ],
      },
    },
  },
  "tags": [
    {
      "name": "TestDiscriminator",
    },
  ],
}
`;
